### PR TITLE
[GUI.performer] Add remove elements function for LineCollisionModel

### PR DESCRIPTION
**Pull request by ScheiklP**
As discussed with @epernod  and @hugtalbot ,
implemented method in `TopologicalChangeManager` to handle `LineCollisionModel` in `TopologicalChangeManager::removeItemsFromCollisionModel`.

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by ScheiklP** [comment]
Oh and here's a minimal scene to test the new feature:

```python
import Sofa
import Sofa.Core


def createScene(
    root_node: Sofa.Core.Node,
) -> Sofa.Core.Node:

    plugin_list = [
        "Sofa.Component.Collision.Detection.Algorithm",  # [BVHNarrowPhase, BruteForceBroadPhase, DefaultPipeline]
        "Sofa.Component.Collision.Detection.Intersection",  # [NewProximityIntersection]
        "Sofa.Component.Collision.Geometry",  # [SphereCollisionModel]
        "Sofa.Component.Collision.Response.Contact",  # [DefaultContactManager]
        "Sofa.Component.Constraint.Projective",  # [FixedConstraint]
        "Sofa.Component.LinearSolver.Iterative",  # [CGLinearSolver]
        "Sofa.Component.Mass",  # [UniformMass]
        "Sofa.Component.ODESolver.Backward",  # [EulerImplicitSolver]
        "Sofa.Component.SolidMechanics.Spring",  # [StiffSpringForceField]
        "Sofa.Component.Topology.Container.Dynamic",  # [EdgeSetTopologyContainer, EdgeSetTopologyModifier]
        "Sofa.Component.Visual",  # [VisualStyle]
        "SofaCarving",  # [CarvingManager]
        "Sofa.Component.StateContainer",  # [MechanicalObject
    ]

    plugin_node = root_node.addChild("Plugins")

    for plugin in plugin_list:
        plugin_node.addObject("RequiredPlugin", pluginName=plugin, name=plugin)

    root_node.addObject("DefaultAnimationLoop")
    root_node.addObject("DefaultVisualManagerLoop")
    root_node.addObject(
        "VisualStyle",
        displayFlags=["showVisual", "showForceFields", "showCollisionModels", "showBehaviorModels", "showInteractionForceFields"],
    )

    root_node.addObject("DefaultPipeline")
    root_node.addObject("BruteForceBroadPhase")
    root_node.addObject("BVHNarrowPhase")
    root_node.addObject("DefaultContactManager", response="PenalityContactForceField")

    root_node.addObject(
        "NewProximityIntersection",
        alarmDistance=3.0,
        contactDistance=0.5,
    )

    root_node.gravity = [0.0, -918, 0.0]

    scene_node = root_node.addChild("scene")
    #############
    # Cutting Sphere
    #############
    cutting_sphere_node = scene_node.addChild("cutting_sphere")
    cutting_sphere_node.addObject("CGLinearSolver")
    cutting_sphere_node.addObject("EulerImplicitSolver")
    cutting_sphere_node.addObject("PointSetTopologyContainer")
    cutting_sphere_node.addObject("MechanicalObject", template="Rigid3d", position=[5.0, 5.0, 0.0, 0.0, 0.0, 0.0, 1.0], showObject=False, showObjectScale=1.0)
    cutting_sphere_node.addObject("UniformMass", totalMass=0.001)
    cutting_sphere_node.addObject("SphereCollisionModel", radius=0.5, group=1, tags="CarvingTool")

    #####################
    # Cuttable EdgeSet Object
    #####################
    edge_node = scene_node.addChild("edges")

    edge_node.addObject("CGLinearSolver")
    edge_node.addObject("EulerImplicitSolver")

    positions = [[x, 0.0, 0.0] for x in range(0, 20)]
    edges = [[x, x + 1] for x in range(len(positions) - 1)]
    edge_node.addObject("EdgeSetTopologyContainer", position=positions, edges=edges)
    edge_node.addObject("EdgeSetTopologyModifier")

    edge_node.addObject("MechanicalObject", showObject=True, showObjectScale=2.0)
    springs = [[first, second, 100, 0.5, 0] for first, second in edges]
    edge_node.addObject("StiffSpringForceField", spring=springs)
    edge_node.addObject("FixedConstraint", indices=[0, len(positions) - 1])
    edge_node.addObject("LineCollisionModel", tags="CarvingSurface")

    root_node.addObject(
        "CarvingManager",
        carvingDistance=0.5,
        active=True,
    )
```

**Comment by ScheiklP** [comment]
@epernod @hugtalbot I just noticed there are problems with this addition, when more than one Sphere is in collision.

If you add another position to the sphere
```
    position = [[5.0, 5.0, 0.0, 0.0, 0.0, 0.0, 1.0], [5.5, 5.0, 0.0, 0.0, 0.0, 0.0, 1.0]]
    cutting_sphere_node.addObject("MechanicalObject", template="Rigid3d", position=position, showObject=False, showObjectScale=1.0)
```
Index 5 is removed twice!
First, 5 is removed and the indices are reordered, so that 5 is now actually the one fixed by the FixedConstraint.
Then, 5 is removed again.

In a more extreme case, this results in a Segfault.
If there are more collision models than positions in the MO to be cut.
```
    position = [[5.0, 5.0, 0.0, 0.0, 0.0, 0.0, 1.0]]*11
    cutting_sphere_node.addObject("MechanicalObject", template="Rigid3d", position=position, showObject=False, showObjectScale=1.0)
```

**EDIT:**
Fixed in the new commit

File: applications/plugins/SofaCarving/CarvingManager.cpp [hunk 1]
```diff
@@ -97,8 +97,6 @@ void CarvingManager::init()
         {
             core::CollisionModel* m = models[i];
             m->getContext()->get(topoMapping);
-            if (topoMapping == NULL) continue;
```
Comment (epernod) [inline applications/plugins/SofaCarving/CarvingManager.cpp]: 2. could you remove this change from this PR not to mix the changes. A proper PR to update the CarvingManager is indeed needed.

**Review by epernod — COMMENTED** [review]
Hello @ScheiklP ,

thank you for this PR. Two points:
1. Could you explain why you have 2x the index 5 in your example?
   Is it because you are using several collision models and the index 5 is added twice from the carvingManager?
   
In the case of segfault, is it because the indice 5 is still to be removed but your vertex buffer last index is now 4 ?

**Comment by ScheiklP** [comment]
> Hello @ScheiklP ,
> 
> thank you for this PR. Two points:
> 
> 1. Could you explain why you have 2x the index 5 in your example?
>    Is it because you are using several collision models and the index 5 is added twice from the carvingManager?
> 
> In the case of segfault, is it because the indice 5 is still to be removed but your vertex buffer last index is now 4 ?

Hi @epernod ,
The collision pipeline correctly detected collisions between index 5 on the LineCollisionModel and all indices on the SphereCollisionModels. In the other removeElement functions, the indices are put into a set to filter duplicates. In the first commit for the new features this was not the case -> 5 was in the indices list multiple times.
The new commit filters out the duplicates.
It should also quite a bit faster, because it does not need to create a set.
Just one sort at creation.

**Comment by epernod** [comment]
Sorry, my question was more, do you know why the index 5 was inserted multiple time?

File: Sofa/GUI/Component/src/sofa/gui/component/performer/TopologicalChangeManager.h [hunk 2]
```diff
@@ -136,6 +137,7 @@ class SOFA_GUI_COMPONENT_API TopologicalChangeManager
 
     Index removeItemsFromTriangleModel(sofa::component::collision::geometry::TriangleCollisionModel<sofa::defaulttype::Vec3Types>* model, const type::vector<Index>& indices) const;
     Index removeItemsFromPointModel(sofa::component::collision::geometry::PointCollisionModel<sofa::defaulttype::Vec3Types>* model, const type::vector<Index>& indices) const;
+    Index removeItemsFromLineModel(sofa::component::collision::geometry::LineCollisionModel<sofa::defaulttype::Vec3Types>* model, const type::vector<Index>& indices) const;
```
Comment (epernod) [inline Sofa/GUI/Component/src/sofa/gui/component/performer/TopologicalChangeManager.h:143]: I know this is not done for the rest of the header, but could you add a small doxygen doc like:
```
/** \brief Method to remove topological elements from a Topology linked to a Line collision model. Only Edge Topology  is supported.
*  \param indices : list of element indices to remove (unique check is done internally)
*/
```

File: Sofa/GUI/Component/src/sofa/gui/component/performer/TopologicalChangeManager.cpp [hunk 3]
```diff
@@ -290,6 +290,35 @@ Index TopologicalChangeManager::removeItemsFromPointModel(PointCollisionModel<so
...
+
+    // copy indices to have a mutable version of the vector
+    type::vector<Index> unique_indices = indices;
+    // sort followed by unique, to remove all duplicates
+    std::sort(unique_indices.begin(), unique_indices.end());
+    unique_indices.erase( std::unique(unique_indices.begin(), unique_indices.end()), unique_indices.end());
+
+    simulation::Node *node_curr = dynamic_cast<simulation::Node*>(topo_curr->getContext());
+    sofa::core::topology::TopologyModifier* topoMod;
+    topo_curr->getContext()->get(topoMod);
```
Comment (damienmarchal) [inline Sofa/GUI/Component/src/sofa/gui/component/performer/TopologicalChangeManager.cpp]: Is such syntax possible ?
```cpp
auto topoMod = topo_curr->getContext()->get<sofa::core::topology::TopologyModifier>() 
```

Comment (epernod) [reply Sofa/GUI/Component/src/sofa/gui/component/performer/TopologicalChangeManager.cpp]: Also, if no modifier is providen in the scene, topoMod will be nullptr.
You should add a check on the pointer @ScheiklP 
Thanks @damienmarchal for the peer review.

**Comment by ScheiklP** [comment]
> Sorry, my question was more, do you know why the index 5 was inserted multiple time?

Ah, yes, because of:
https://github.com/sofa-framework/sofa/blob/master/applications/plugins/SofaCarving/CarvingManager.cpp#L183-L194

We could change `elemsToRemove` from `vector` to `set` at the cost of a bit performance.

File: Sofa/GUI/Component/src/sofa/gui/component/performer/TopologicalChangeManager.cpp [hunk 4]
```diff
@@ -290,6 +290,37 @@ Index TopologicalChangeManager::removeItemsFromPointModel(PointCollisionModel<so
...
+    type::vector<Index> unique_indices = indices;
+    // sort followed by unique, to remove all duplicates
+    std::sort(unique_indices.begin(), unique_indices.end());
+    unique_indices.erase(std::unique(unique_indices.begin(), unique_indices.end()), unique_indices.end());
+
+    auto topo_mod = topo_curr->getContext()->get<sofa::core::topology::TopologyModifier>();
+    if(dynamic_cast<EdgeSetTopologyModifier*>(topo_mod) == nullptr){
+        msg_warning("TopologicalChangeManager") << "Cannot find an EdgeSetTopologyModifier to perform the changes.";
+        return 0;
+    }
```
Comment (alxbilger) [inline Sofa/GUI/Component/src/sofa/gui/component/performer/TopologicalChangeManager.cpp]:
```suggestion
    auto topo_mod = topo_curr->getContext()->get<EdgeSetTopologyModifier>();
    if(topo_mod == nullptr)
    {
        msg_warning("TopologicalChangeManager") << "Cannot find an EdgeSetTopologyModifier to perform the changes.";
        return 0;
    }
```

**Comment by damienmarchal** [comment]
Reviewed, ready when the CI is back online.

**Comment by damienmarchal** [comment]
[ci-build]

**Comment by damienmarchal** [comment]
No compiling anymore.

**Comment by fredroy** [comment]
[ci-build][with-all-tests]

**Pull request merged**